src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingSourceGenerator.Parser.cs

```diff
@@ -922,29 +964,59 @@ public static bool ContainsGenericParameters(INamedTypeSymbol type)
                 return false;
             }
 
-            private static bool HasPublicParameterlessCtor(INamedTypeSymbol type)
+            private static bool HasPublicParameterLessCtor(INamedTypeSymbol type) =>
+                CannotBeConstructed(type)
+                ? false
+                : type.InstanceConstructors.SingleOrDefault(ctor => ctor.DeclaredAccessibility is Accessibility.Public && ctor.Parameters.Length is 0)
+                    is not null;
+
+            private static IMethodSymbol? GetConstructor(INamedTypeSymbol type)
             {
-                if (type.IsAbstract || type.TypeKind == TypeKind.Interface)
+                if (type.Name is "ReadonlyRecordStructTypeOptions")
                 {
-                    return false;
                 }
 
-                if (type is not INamedTypeSymbol namedType)
+                if (CannotBeConstructed(type))
                 {
-                    return false;
+                    return null;
                 }
 
-                foreach (IMethodSymbol ctor in namedType.InstanceConstructors)
+                IMethodSymbol? publicParameterlessCtor = null;
+                IMethodSymbol? publicParameterizedCtor = null;
+                bool hasMultiplePublicParameterizedCtors = false;
+
+                foreach (IMethodSymbol ctor in type.InstanceConstructors)
                 {
-                    if (ctor.DeclaredAccessibility == Accessibility.Public && ctor.Parameters.Length == 0)
+                    if (ctor.DeclaredAccessibility is not Accessibility.Public)
                     {
-                        return true;
+                        continue;
+                    }
+
+                    if (ctor.Parameters.Length is 0)
+                    {
+                        publicParameterlessCtor = ctor;
```

Comment:
> publicParameterlessCtor = ctor; [](http://example.com/codeflow?start=23&length=32)

If you find the public parameterless constructor, why you don't just return it?

Reply:
Refactored a bit. Impl is the same as reflection. Reflection prefers a single parameterized ctor over a parameterless one.